### ISSUE TITLE
Android: black screen on resume and no response

### ISSUE THREAD

**Issue by pbienst** [koreader/koreader#10531]
* KOReader version: 2023.05
* Device: Android POCO XT G4

When resuming the Android app, I sometimes get a black screen and the app does not react anymore.

Killing the app through the carousel and restarting it solves the problem. 

In the attached log file, the statement 'cannot blit: no window' seems suspicious.

[crash.log](https://github.com/koreader/koreader/files/11642065/crash.log)

**Comment by pazos** [comment]
> In the attached log file, the statement 'cannot blit: no window' seems suspicious

You can usually ignore these warning messages unless you see two or more in a row.

**Comment by pbienst** [comment]
I'm more than happy to ignore these warnings, but the fact is that screen stays black :-)

**Comment by pbienst** [comment]
BTW, I see you've labelled this bug report as 'need more info'. I'm happy to provide whatever extra info you think is valuable beyond the crash log, but you need to let me know what you want from me :-)

**Comment by NiLuJe** [comment]
Verbose debug lkogs would be a start ;).

**Comment by pazos** [comment]
What @NiLuJe said.

Also:

- can you reproduce on previous versions? (ie: v2023.04/v2023.03)
- can you install `pidcat` and get the full log with `pidcat org.koreader.launcher`?


> I'm more than happy to ignore these warnings, but the fact is that the screen stays black :-)

Yup ;)

My statement was meant to say these messages *also* happen on devices without black screens.

**Comment by pbienst** [comment]
Here's the extended log:

[crash.log](https://github.com/koreader/koreader/files/11649645/crash.log)

This is the first version of koreader I'm using on my new phone, so I don't know what the previous versions did.

Just let me know if the attached log is not enough, then I can try to go the pidcat route. It won't be easy though, as the error is not reproducible and I wont' be wired to my desktop all the time...

**Comment by pbienst** [comment]
Here's a log collected through adb:

[adb_log.txt](https://github.com/koreader/koreader/files/11780850/adb_log.txt)

(To avoid this bug getting ignored, please remove the 'need more info' tag :-) )

**Comment by NiLuJe** [comment]
> (To avoid this bug getting accidentally ignored, please remove the 'need more info' tag :-) )

Rest assured, the "Android" tag is enough for that ;p.

**Comment by Frenzie** [comment]
Removed, but I think I mainly look at the issue title. As such also edited slightly, because "black screen on resume" makes me think that's slightly inconvenient but no big deal. ;-)

**Comment by NiLuJe** [comment]
> Here's a log collected through adb:

Since there's only 30s worth of data in there, can you pinpoint the timestamp at which you resumed?

(I'm pretty sure we'll actually want the log from when you/Android *suspended*, as that's liable to be more interesting).

TL;DR: Moar logz!

**Comment by pbienst** [comment]
These 30 secs should contain everything, from putting koreader in the background in the beginning, to resuming it at the end of the log.

**Comment by pazos** [comment]
> (To avoid this bug getting accidentally ignored, please remove the 'need more info' tag :-) )

I'm fairly sure that's not how the things work here :)

For one part **I** would need **more info**. As in:

> can you reproduce on previous versions? (ie: v2023.04/v2023.03)

Please try, and if you can reproduce:

> can you install pidcat and get the full log with pidcat org.koreader.launcher?

Thanks!

**Comment by pbienst** [comment]
It took me a couple of hours to set up Koreader the way I like, and I cannot seem downgrade without uninstalling. Will this reset my settings?

**Comment by pazos** [comment]
> Will this reset my settings?

Nope. App settings are placed in `/sdcard/koreader`.  You should be able to switch between versions back and forth without losing data. Anyhow, a backup is recommended if you're going to go to a much older release and it really never hurts so you can backup `/sdcard/koreader` to `/sdcard/koreader.bak`, then downgrade and see if you can reproduce. Once everything went well you can remove the backup.

**Comment by pbienst** [comment]
The April version has the same problem. To what extent do you expect the logcat for this version to be different from the one I posted earlier? I.e., is this information really required? I'm away from my computer for the next few weeks...

**Comment by pbienst** [comment]
Here's a log file for the April version, where I did lots of switching between apps, and finally got a black screen in the end:

[koreader log 2.txt](https://github.com/koreader/koreader/files/12145825/koreader.log.2.txt)

**Comment by pazos** [comment]
Hi. The logs contain a whole bunch of unrelated messages.

Next time just use `logcat` to filter those messages or the [full adb command](https://github.com/koreader/koreader/blob/master/kodev#L1176)

It's also possible to get a filtered log in "help" -> "report a bug". Makes way easier for us to grasp what happens.

Anyhow, the logs contains no errors on our end. All the activity lifecycle steps are there and catched correctly.

Not sure if this is a bug here or on your phone. In the mean time I would try to disable battery optimization for koreader (it is a setting buried on android settings but there's a link to it on "settings" -> "device" in KO)

**Comment by pbienst** [comment]
Sorry, I thought I filtered the adb log on 'koreader', but perhaps I forgot to press enter :-)

I tried turning off the battery optimisation, but I'm afraid that didn't make a difference...

**Comment by rtega** [comment]
I'm also seeing this since years on android on two different devices. I turned of battery optimisation but that doesn't change a thing. What else can we try since this is terribly annoying?

**Comment by pbienst** [comment]
Just a tidbit of extra information: before the screen turns black after context switching, the text is visible for a split-second in my case.

**Comment by rtega** [comment]
Indeed, I usually also see that.

**Comment by NiLuJe** [comment]
> What else can we try since this is terribly annoying?

Providing the requested logs would be a start ;).

**Comment by rtega** [comment]
[koreader-logcat.log](https://github.com/koreader/koreader/files/12898389/koreader-logcat.log)
Does this make you guys any wiser? I did a full adb command as requested and removed everything that was prior to my starting the koreader process.

**Comment by rtega** [comment]
Did anybody find time to look at this? Do you need other logs?

**Comment by pazos** [comment]
> Did anybody find time to look at this? Do you need other logs?

It is an ANR, same underlying cause as in #11073

**Comment by pbienst** [comment]
I'm afraid nothing has changed... Could you please reopen this bug?

**Comment by rtega** [comment]
I concur. Nothing has changed at my at either. Turning of partial rendering does nothing.

**Comment by pazos** [comment]
@pbienst @rtega:  having a bunch of duplicate tickets won't help to get them fixed. It is even worse as people tend to comment on a subset of those and ignore the other ones, so the info gets scattered into different places.

tl;dr: use #11703 or show (with reproducible steps and paired logs) that the issue you're experiencing here is not caused by the same ANR the other issue covers.

**Comment by pbienst** [comment]
Sure, here's three arguments:

- #11703 seems like a completely different issue with a completely different set of symptoms. There, it's about changing font size and the app becoming unresponsive while the screen still shows text. Here, it's about the screen turning completely black a split second after resuming the app. 
- In my logs, there are never lines like those in #11703, that indicate an ANR: 
```
11-04 11:35:21.210 1516 1559 V KOReader: _stepRerenderingAutomation( nil ), currently 1
11-04 11:35:22.212 1516 1559 V KOReader: _stepRerenderingAutomation( nil ), currently 1
```
- The fact that these are two different issues seems confirmed by the fact that the fix for one issue did not fix the second issue :-)

We really appreciate your efforts in fixing Android bugs, but unfortunately, it seems like fixing #11703 did not fix this bug...

**Comment by pazos** [comment]
@pbienst:

1. #11703 isn't fixed. This is a duplicate (see @rtega report, for instance)
2. Your issues cannot be reproduced. You failed to provide logs and repro steps
3. Even with logs and repro steps it cannot be fixed if it can't be reproduced on any other device.

I'm trying to keep an eye on the entire KOReader on android thingy. I have no problems keeping a bunch of tickets opened as long as they have some context, some reproducible steps and the relevant logs.

Feel free to open a new ticket for your particular issue if you fill the void. Also, feel free to try to reproduce the issue you're experiencing in any other device and/or fw version.

**Comment by rtega** [comment]
Sorry to interject:

> @pbienst:
> 
> 1. #11703 isn't fixed. This is a duplicate (see @rtega report, for instance)
I see no mention of a black screen on that bug so I wonder why this should be the same issue. What led you to that conclusion? In any case, this issue is not solved by turning off partial rendering at my end. I should mention that my phone is also a Samsung (model S10+).

> 2. Your issues cannot be reproduced. You failed to provide logs and repro steps
 I provided the logs months ago. See above.

> 3. Even with logs and repro steps it cannot be fixed if it can't be reproduced on any other device.
 It is an issue that pops up once an hour or something so I dunno how you can reliably reproduce it. 
 
> I'm trying to keep an eye on the entire KOReader on android thingy. I have no problems keeping a bunch of tickets opened as long as they have some context, some reproducible steps and the relevant logs.
All respect for you devs. We are just trying to get an issue solved here.

> 
> Feel free to open a new ticket for your particular issue if you fill the void. Also, feel free to try to reproduce the issue you're experiencing in any other device and/or fw version.

**Comment by pbienst** [comment]
OK, I missed that #11703 was a duplicate of an unsolved bug, sorry. 

I looked at the other linked reports, but they also seem very different in terms of symptoms. In our case, the whole system does not become unresponsive like in an ANR. The screen turns black after resuming from sleep (I don't think any of the other reports mention this), while the Android OS stays responsive.

With respect to logs, I have several logs at the start of the report, and rtega also provided logs, who sees the issue on two different devices.

Since the issue appears at random times, it is unfortunately not deterministically reproducible. I realise this makes the problem infinitely more difficult to study. This is of course very frustrating, but that is not a reason to brush it under the carpet by closing this bug report...

**Comment by NiLuJe** [comment]
More logs are always helpful, don't assume because someone (or even you yourself) provided logs in the past means that newer logs wouldn't be helpful.

It may not, but more data certainly never hurts.

**Comment by pazos** [comment]
As far as I am concerned all bugs are ANRs until proven otherwise.

That's what happens here.

1. @pbienst opens a ticket
2. ticket need more info
3. @rtega provides more info with an ANR

It doesn't matter if sympthoms match or not. ANRs on Android/KOReader have been documented for a while. Most of them addressed by inhibiting input. 

I'm not going to repeat myself in each new ticket but you can look yourself in https://github.com/search?q=repo%3Akoreader%2Fkoreader+ANR&type=issues&p=1

The only **real** reason to have this ticket reopened is a proper log by @pbienst that reproduces a crash and **does not** contain an ANR.

**Comment by pbienst** [comment]
Just out of curiosity, which part of my own log from June 18 points to an ANR?
